Initialize Caffe2 only when running Caffe2 benchmarks

ilia-cher opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #20157 Support for Eigen thread pool
* #20087 Native ATen/Parallel backend
* #20057 Split ATen/Parallel into interface and backend
* #20050 Move inter-op settings into ATen/Parallel
* #20043 Port ATen/native to ATen/Parallel
* #20032 Port THNN to ATen/Parallel
* #20002 Remove explicit checks for parallelism from TH
* #19105 Port TH library to ATen/Parallel
* #19997 Intra-op parallel microbenchmarks for PT
* #19993 Support for non-contiguous tensors and arbitrary dtypes in PT benchmarks
* **#19980 Initialize Caffe2 only when running Caffe2 benchmarks**
* #19749 Allow a non-OpenMP based build

Summary:
At the moment Caffe2's GlobalInit conflicts with PyTorch
at::init_num_threads;

In this PR we initialize Caffe2 only when necessary - when Caffe2
benchmarks are requested (by default)

Test Plan:
BUILD_BINARY=1 BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1 python setup.py develop --cmake
cd benchmarks
python -m operator_benchmark.ops.matmul_test --framework PyTorch

Differential Revision: [D15229797](https://our.internmc.facebook.com/intern/diff/D15229797)

Commit:
Initialize Caffe2 only when running Caffe2 benchmarks

Summary:
At the moment Caffe2's GlobalInit conflicts with PyTorch
at::init_num_threads;

In this PR we initialize Caffe2 only when necessary - when Caffe2
benchmarks are requested (by default)

Test Plan:
BUILD_BINARY=1 BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1 python setup.py develop --cmake
cd benchmarks
python -m operator_benchmark.ops.matmul_test --framework PyTorch

Commit:
Update on "Initialize Caffe2 only when running Caffe2 benchmarks"

Initialize Caffe2 only when running Caffe2 benchmarks

Summary:
At the moment Caffe2's GlobalInit conflicts with PyTorch
at::init_num_threads;

In this PR we initialize Caffe2 only when necessary - when Caffe2
benchmarks are requested (by default)

Test Plan:
BUILD_BINARY=1 BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1 python setup.py develop --cmake
cd benchmarks
python -m operator_benchmark.ops.matmul_test --framework PyTorch

gh-metadata: pytorch pytorch 19980 gh/ilia-cher/6/head

Commit:
Update on "Initialize Caffe2 only when running Caffe2 benchmarks"

Initialize Caffe2 only when running Caffe2 benchmarks

Summary:
At the moment Caffe2's GlobalInit conflicts with PyTorch
at::init_num_threads;

In this PR we initialize Caffe2 only when necessary - when Caffe2
benchmarks are requested (by default)

Test Plan:
BUILD_BINARY=1 BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1 python setup.py develop --cmake
cd benchmarks
python -m operator_benchmark.ops.matmul_test --framework PyTorch

gh-metadata: pytorch pytorch 19980 gh/ilia-cher/6/head

Commit:
Update on "Initialize Caffe2 only when running Caffe2 benchmarks"

Initialize Caffe2 only when running Caffe2 benchmarks

Summary:
At the moment Caffe2's GlobalInit conflicts with PyTorch
at::init_num_threads;

In this PR we initialize Caffe2 only when necessary - when Caffe2
benchmarks are requested (by default)

Test Plan:
BUILD_BINARY=1 BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1 python setup.py develop --cmake
cd benchmarks
python -m operator_benchmark.ops.matmul_test --framework PyTorch

gh-metadata: pytorch pytorch 19980 gh/ilia-cher/6/head

Commit:
Update on "Initialize Caffe2 only when running Caffe2 benchmarks"

Initialize Caffe2 only when running Caffe2 benchmarks

Summary:
At the moment Caffe2's GlobalInit conflicts with PyTorch
at::init_num_threads;

In this PR we initialize Caffe2 only when necessary - when Caffe2
benchmarks are requested (by default)

Test Plan:
BUILD_BINARY=1 BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1 python setup.py develop --cmake
cd benchmarks
python -m operator_benchmark.ops.matmul_test --framework PyTorch

gh-metadata: pytorch pytorch 19980 gh/ilia-cher/6/head

dzhulgakov commented on benchmarks/operator_benchmark/benchmark_core.py:
```diff
@@ -140,8 +140,10 @@ def run(self):
-                continue
+            if self.args.framework:
+                frameworks = benchmark_utils.get_requested_frameworks(self.args.framework)
+                if all([fr not in full_test_id for fr in frameworks]):
```
nit: [] is redundant

dzhulgakov commented on benchmarks/operator_benchmark/benchmark_runner.py:
```diff
@@ -73,13 +73,14 @@ def main():
-        help='Run PyTorch or Caffe2 operators',
-        default=None)
+        help='Comma-delimited list of frameworks to test (Caffe2, PyTorch)',
+        default="Caffe2,PyTorch")
```
I think you can put `action=append` here and not rely on comma-separated list

ilia-cher replied:
not sure which one you mean, might be appendTuple, though comma-separated list of also pretty standard way to pass args

ilia-cher replied:
though in this case we'll have to use `--framework` multiple times, I think comma-separated list is more standard and compact

dzhulgakov reviewed (approved):
minor suggestions, but looks good

dzhulgakov commented:
And we should really unify initialization (also cc @smessmer). Would be nice to create an issue/quip with some plan about it

Commit:
Update on "Initialize Caffe2 only when running Caffe2 benchmarks"

Initialize Caffe2 only when running Caffe2 benchmarks

Summary:
At the moment Caffe2's GlobalInit conflicts with PyTorch
at::init_num_threads;

In this PR we initialize Caffe2 only when necessary - when Caffe2
benchmarks are requested (by default)

Test Plan:
BUILD_BINARY=1 BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1 python setup.py develop --cmake
cd benchmarks
python -m operator_benchmark.ops.matmul_test --framework PyTorch

gh-metadata: pytorch pytorch 19980 gh/ilia-cher/6/head

Commit:
Update on "Initialize Caffe2 only when running Caffe2 benchmarks"

Initialize Caffe2 only when running Caffe2 benchmarks

Summary:
At the moment Caffe2's GlobalInit conflicts with PyTorch
at::init_num_threads;

In this PR we initialize Caffe2 only when necessary - when Caffe2
benchmarks are requested (by default)

Test Plan:
BUILD_BINARY=1 BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1 python setup.py develop --cmake
cd benchmarks
python -m operator_benchmark.ops.matmul_test --framework PyTorch

gh-metadata: pytorch pytorch 19980 gh/ilia-cher/6/head

Commit:
Update on "Initialize Caffe2 only when running Caffe2 benchmarks"

Initialize Caffe2 only when running Caffe2 benchmarks

Summary:
At the moment Caffe2's GlobalInit conflicts with PyTorch
at::init_num_threads;

In this PR we initialize Caffe2 only when necessary - when Caffe2
benchmarks are requested (by default)

Test Plan:
BUILD_BINARY=1 BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1 python setup.py develop --cmake
cd benchmarks
python -m operator_benchmark.ops.matmul_test --framework PyTorch

gh-metadata: pytorch pytorch 19980 gh/ilia-cher/6/head

ilia-cher commented:
Created https://github.com/pytorch/pytorch/issues/20189

Commit:
Update on "Initialize Caffe2 only when running Caffe2 benchmarks"

Initialize Caffe2 only when running Caffe2 benchmarks

Summary:
At the moment Caffe2's GlobalInit conflicts with PyTorch
at::init_num_threads;

In this PR we initialize Caffe2 only when necessary - when Caffe2
benchmarks are requested (by default)

Test Plan:
BUILD_BINARY=1 BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1 python setup.py develop --cmake
cd benchmarks
python -m operator_benchmark.ops.matmul_test --framework PyTorch

gh-metadata: pytorch pytorch 19980 gh/ilia-cher/6/head

Commit:
Update on "Initialize Caffe2 only when running Caffe2 benchmarks"

Initialize Caffe2 only when running Caffe2 benchmarks

Summary:
At the moment Caffe2's GlobalInit conflicts with PyTorch
at::init_num_threads;

In this PR we initialize Caffe2 only when necessary - when Caffe2
benchmarks are requested (by default)

Test Plan:
BUILD_BINARY=1 BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1 python setup.py develop --cmake
cd benchmarks
python -m operator_benchmark.ops.matmul_test --framework PyTorch

gh-metadata: pytorch pytorch 19980 gh/ilia-cher/6/head

Commit:
Update on "Initialize Caffe2 only when running Caffe2 benchmarks"

Initialize Caffe2 only when running Caffe2 benchmarks

Summary:
At the moment Caffe2's GlobalInit conflicts with PyTorch
at::init_num_threads;

In this PR we initialize Caffe2 only when necessary - when Caffe2
benchmarks are requested (by default)

Test Plan:
BUILD_BINARY=1 BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1 python setup.py develop --cmake
cd benchmarks
python -m operator_benchmark.ops.matmul_test --framework PyTorch

gh-metadata: pytorch pytorch 19980 gh/ilia-cher/6/head
